File: src/198.house_robber.kt
import kotlin.math.max

fun main() {
    println(rob(intArrayOf(1, 2, 3, 1)))
    println(rob(intArrayOf(2, 7, 9, 3, 1)))
    println(rob(intArrayOf(2, 1, 1, 2)))
}

/// todo
fun rob(nums: IntArray): Int {
    val dp = nums.toMutableList()

    for (i in nums.indices) {
        for (j in i + 2..<nums.size) {
            dp[j] = max(dp[j], dp[i] + nums[j])
        }
    }

    return dp.max()
}

File: lib/198.house_robber.dart
import 'dart:math';

void main() {
  print(rob([1, 2, 3, 1]));
  print(rob([2, 7, 9, 3, 1]));
  print(rob([2, 1, 1, 2]));
}

/// todo
int rob(List<int> nums) {
  final dp = List<int>.from(nums);

  for (var i = 0; i < nums.length; i++) {
    for (var j = i + 2; j < nums.length; j++) {
      dp[j] = max(dp[j], dp[i] + nums[j]);
    }
  }

  return dp.isEmpty ? 0 : dp.reduce(max);
}
